common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala

```diff
@@ -38,15 +53,188 @@ trait MessageConsumer {
+
+    require(consumer.maxPeek <= maxPipelineDepth, "consumer may not yield more messages per peek than permitted by max depth")
+
+    private val outstandingMessageQueue = new ConcurrentLinkedQueue[(String, Int, Long, Array[Byte])]()
+    private val outstandingMessageCount = new AtomicInteger() // avoid non-contant time size calculation on queue
+    private val handlerCapacity = new AtomicInteger(maximumHandlerCapacity)
+
+    private implicit val tid = TransactionId.dispatcher
+
+    logging.info(this, s"handler capacity = $maximumHandlerCapacity, pipeline fill at = $pipelineFillThreshold, pipeline depth = $maxPipelineDepth")
+    startWith(Uninitialized, MessageFeed.NoData)
+
+    when(Uninitialized) {
+        case Event(Ready, _) =>
+            fillPipeline()
+            goto(FillingPipeline)
+
+        case _ => stay
+    }
+
+    // wait for fill to complete, and keep filling if there is
+    // capacity otherwise wait to drain
+    when(FillingPipeline) {
+        case Event(Processed, _) =>
+            val capacity = updateHandlerCapacity()
+            sendOutstandingMessages(capacity)
+            stay
+
+        case Event(FillCompleted, _) =>
+            if (shouldFillQueue()) {
+                fillPipeline()
+                stay
+            } else {
+                goto(DrainingPipeline)
+            }
+
+        case _ => stay
+    }
+
+    when(DrainingPipeline) {
+        case Event(Processed, _) =>
+            val capacity = updateHandlerCapacity()
+            sendOutstandingMessages(capacity)
+            if (shouldFillQueue()) {
+                fillPipeline()
+                goto(FillingPipeline)
+            } else stay
+
+        case _ => stay
+    }
+
+    if (autoStart) self ! Ready
+
+    private implicit val ec = context.system.dispatcher
+
+    private def fillPipeline(): Unit = Future {
+        if (outstandingMessageCount.get <= pipelineFillThreshold) {
+            Try {
+                // Grab next batch of messages and commit offsets immediately
+                // essentially marking the activation as having satisfied "at most once"
+                // semantics (this is the point at which the activation is considered started).
+                // If the commit fails, then messages peeked are peeked again on the next poll.
+                // While the commit is synchronous and will block until it completes, at steady
+                // state with enough buffering (i.e., maxPipelineDepth > maxPeek), the latency
+                // of the commit should be masked.
+                val records = consumer.peek(longPollDuration)
+                consumer.commit()
+                (records, records.size)
+            }.map {
+                case (records, count) =>
+                    outstandingMessageQueue.addAll(records)
+                    outstandingMessageCount.addAndGet(count)
+                    sendOutstandingMessages(handlerCapacity.get)
```

Comment:
These should happen responding to an even being sent by the Future (the future could send: `Seq[KafkaMessage]`)